src/QsCompiler/Transformations/FunctorGeneration.cs

```diff
@@ -148,6 +152,123 @@ public override QsStatementKind OnConjugation(QsConjugation stm)
         }
     }
 
+    /// <summary>
+    /// Scope transformation that splits any nested operation calls into separate statements so
+    /// that they can be properly reversed. This is necessary to avoid out of order execution of the
+    /// automatically generated adjoint. It is safe to do because an adjointable operation must return
+    /// Unit, so any nested calls can be replaced by Unit and those calls moved to separate,
+    /// ordered statements.
+    /// </summary>
+    internal class ExtractNestedOperationCalls
+    : SyntaxTreeTransformation<ExtractNestedOperationCalls.TransformationsState>
+    {
+        internal class TransformationsState
+        {
+            /// <summary>
+            /// Accumulates statements that have been lifted from the current statement.
+            /// </summary>
+            public List<QsStatement> AdditionalStatements = new List<QsStatement>();
+
+            /// <summary>
+            /// Tracks the current expression that is being evaluated, keeping previous
+            /// expressions in the stack so that we can return to those when leaving a nested
+            /// evaluation.
+            /// </summary>
+            public Stack<TypedExpression> CurrentExpression = new Stack<TypedExpression>();
+
+            /// <summary>
+            /// Allows us to remember the current statement location, and use that for the generated
+            /// statements that get added when extracting nested expressions.
+            /// </summary>
+            public QsNullable<QsLocation> StatementLocation = QsNullable<QsLocation>.Null;
+        }
+
+        public ExtractNestedOperationCalls()
+        : base(new TransformationsState())
+        {
+            this.Statements = new StatementTransformation(this);
+            this.Expressions = new ExpressionTransformation(this);
+            this.ExpressionKinds = new ExpressionKindTransformation(this);
```

Comment:
Since you are not doing any type transformations, it would be nice to disable them so the transformation doesn't unnecessarily recur into and rebuild all types. 
```suggestion
            this.ExpressionKinds = new ExpressionKindTransformation(this);
            this.Types = new TypeTransformation<TransformationState>(this, TransformationOptions.Disabled);
```